### ISSUE TITLE
Plot classifier space in 2D

### ISSUE THREAD

**Issue by stefanobellelli** [stefanobellelli/nonce3vec#7]
[Some code to start from](https://github.com/minimalparts/Tutorials/blob/master/Translation/utils.py) (see `make_figure` function).

Notes (by [minimalparts](https://github.com/minimalparts)):

> To reduce your data, however, I wouldn't use PCA but t-SNE, directly on a numpy matrix.
> Assuming you already have a numpy array in k x n dimensions, corresponding to your vector space (let's call it mat), the t-SNE code would be:
> ```
> from sklearn.manifold import TSNE`
> def run_TSNE(mat):
> m_2d = TSNE(n_components=2).fit_transform(mat)
> return np.array(m_2d)
> ```
> Then, m_2d is what you feed to matplotlib.

**Comment by stefanobellelli** [comment]
I'm not sure what I should do with the `labels` and `columns` in the `make_figure` fun.

**Comment by stefanobellelli** [comment]
How do I arrange the matrix? Is it just an array of linear arrays (i.e. the vectors)? E.g.:
`np.array([[1,2,0,5],[0,2,3,1],[4,3,3,0]])`?

Also, does this include *all* vectors, or does one have the Wiki space and another the BNC?

**Comment by stefanobellelli** [comment]
Anyway, I've tried with a plain Python nested list (e.g.: `[[1,2,4],[5,5,0],[0,1,3]]`) over a 100-vectors space. This is the result:
![asdfg](https://user-images.githubusercontent.com/24751065/43370562-5eb4b8e4-9381-11e8-8a77-6ae31c65e89a.png)

**Comment by minimalparts** [comment]
This looks good, but have one colour for the Wiki vectors and another for the BNC, so that we see where the classes are. You can use the color argument of plt for that:

    plt.scatter(v[0],v[1],color=mycolor)

About the array... you're passing a numpy array to the t-SNE function and it also returns a numpy array.

**Comment by stefanobellelli** [comment]
Some questions:
1) I'm plotting directly by the method embedded in the DataFrame object, so what should I do with `plt`? Currently I'm using `cm.get_cmap('nipy_spectral')` to get a colour map that I pass to `DataFrame.plot`;
2) the graph above only includes one corpus... should I join both? If so, how do I tell the script which vectors are in which class?;
3) what's the difference between feeding lists and feeding numpy arrays? If I feed a list, it doesn't throw an error, but the result is different.

**Comment by minimalparts** [comment]
Ah yes, sorry, the code was running over pandas... Then you want to use the possible options to plot two dataframes. Check this out: https://stackoverflow.com/questions/46354815/scatter-plotting-data-from-two-different-data-frames-in-python. You basically want one dataframe for the Wiki vectors and one for the BNC vectors. And set colours in the plot function as show on SO.

sklearn will attempt to convert your list to a numpy array: http://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html. I don't exactly know how it does this, but if the result is slightly different from directly feeding the numpy array, the t-SNE may end up looking completely different (although may not be *that* different if you look at actual distances between points in the 2D space).

**Comment by stefanobellelli** [comment]
One thing I don't get is why, if I do
```python3
x = TSNE(n_components = 2).fit_transform(srcmat)
```
then `x` is a numpy array, but if I do
```python3
x = np.array(TSNE(n_components = 2).fit_transform(srcmat))
```
then `x` is still a numpy array, BUT COMPLETELY DIFFERENT.

Is it normal? What do I have to do?

Edit: funnily enough, if I do
```python3
x = TSNE(n_components = 2).fit_transform(srcmat)
y = np.array(x)
```
Then `x` and `y` will be the same.

**Comment by stefanobellelli** [comment]
By the way, I've played a bit with the code (please @minimalparts take a look at the new version and check if everything's in order).
This is the resulting figure (I've used POS unigrams, vector size = 50). Looks pretty awful to me, despite the linear SVM being pretty good ad sorting them out.

![fig](https://user-images.githubusercontent.com/24751065/44555930-bb2b0a00-a737-11e8-924f-0c99c9e6a2c4.png)

**Comment by minimalparts** [comment]
I think we might be talking at cross-purposes. t-SNE uses gradient descent (see the paper: http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf), so if you run your program several times, it will anyway give you different results in terms of the actual components of the vectors it is outputting. What should roughly be the same is the respective distances between points (I say 'roughly' because again, it's a stochastic process).

sklearn will return a numpy array, so the reason why you get the same array for 

    x = TSNE(n_components = 2).fit_transform(srcmat)
    y = np.array(x)

is that your second statement doesn't do anything.

With regard to the plot, actually, it is exactly what I would expect for POS unigrams, so very nice :) As I have said before, I remain puzzled by the fact that they perform so well in the SVM classification. I still think we might have a bug (especially given that they give the same results as the trigrams).

Have you tried plotting for your best configuration? How does it change things?

**Comment by stefanobellelli** [comment]
> I think we might be talking at cross-purposes. t-SNE uses gradient descent (see the paper: http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf), so if you run your program several times, it will anyway give you different results in terms of the actual components of the vectors it is outputting. What should roughly be the same is the respective distances between points (I say 'roughly' because again, it's a stochastic process).

Okay. But they're _very_ different. See:
```
[[ 12.9951935 -25.216873 ]
 [ 49.600315   12.3205   ]
 [ 56.141525  -13.292829 ]
 ...
 [ 68.937996   -7.9838862]
 [-30.162756    7.0588255]
 [  1.5796291 -38.493107 ]]
```
```
[[-11.251794  -11.479982 ]
 [ -3.6287172 -47.90537  ]
 [-21.939651  -60.147755 ]
 ...
 [-13.621424  -70.49706  ]
 [  8.08276    42.98261  ]
 [  6.9075685  -9.0372305]]
```
Does this look ok?



> With regard to the plot, actually, it is exactly what I would expect for POS unigrams, so very nice :) As I have said before, I remain puzzled by the fact that they perform so well in the SVM classification. I still think we might have a bug (especially given that they give the same results as the trigrams).

I've run it on brand new results that I've computed some days ago on the servers. Which you can find at [https://github.com/stefanobellelli/nonce3vec/blob/master/output/results.ods](url).

**Comment by stefanobellelli** [comment]
> Have you tried plotting for your best configuration? How does it change things?

POS bigrams, without exclusions, 500 dimensions:
![fig](https://user-images.githubusercontent.com/24751065/44584224-8d31de00-a7a8-11e8-995b-b3ecb061896c.png)

**Comment by minimalparts** [comment]
About the returned arrays: I think all is fine. The initialisation is random, so you're bound to get very different outputs. The main thing is that similarities in the full space are conserved in the 2D space. To understand this, imagine you have three words in the full space. You can draw a triangle from those three points, and the lengths of the triangle's sides tell you how similar the words are. Now, you could rotate and shift that triangle through space, leading to completely different vectors for each word, and still conserve the similarities (think of the triangle as a rigid body moving through space). So it doesn't matter what the actual coordinates of your vectors are, as long as the distances are correct. 

Okay, the picture for the best config is not great either... Well, shame that there is nothing nice to visually show, but this just goes to show that the SVM really needs those 500 dimensions :) Plus, it's interesting to see that *some* portion of the classes is actually clearly defined, even in 2D (see the bits of blue / red sticking out). We obviously have a few very clear cases in the data. (But which ones?)

**Comment by stefanobellelli** [comment]
So, is the plotter ready to go? Or do we need to add features?

**Comment by minimalparts** [comment]
No, it's great as is. We may want to run it on a few other cases, but let's first wait for the general analysis.
